poc_mobile_verifier/src/verifier.rs

```diff
@@ -0,0 +1,192 @@
+    pub verifier: Verifier,
+}
+
+impl VerifierDaemon {
+    pub async fn run(mut self, shutdown: &triggered::Listener) -> Result {
+        tracing::info!("Starting verifier service");
+
+        let reward_period = Duration::hours(self.reward_period_hours);
+        let verification_period = reward_period / self.verifications_per_period;
+
+        loop {
+            let now = Utc::now();
+            // Maybe name these "epoch_since_*" and "epoch_since_*_duration"
+            let verify_epoch = self.verifier.get_verify_epoch(now);
+            let reward_epoch = self.verifier.get_reward_epoch(now);
+            let verify_epoch_duration = epoch_duration(&verify_epoch);
+            let reward_epoch_duration = epoch_duration(&reward_epoch);
+
+            // If we started up and the last verification epoch was too recent,
+            // we do not want to re-verify.
+            let mut sleep_duration = if verify_epoch_duration >= verification_period
+                // We always want to verify before a reward 
+                || reward_epoch_duration >= reward_period
+            {
+                tracing::info!("Verifying epoch: {:?}", verify_epoch);
+                // Attempt to verify the current epoch:
+                self.verify_epoch(verify_epoch).await?;
+                verification_period
+            } else {
+                verification_period - verify_epoch_duration
+            };
+
+            // If the current duration since the last reward is exceeded, attempt to
+            // submit rewards
+            if reward_epoch_duration >= reward_period {
+                tracing::info!("Rewarding epoch: {:?}", reward_epoch);
+                self.reward_epoch(reward_epoch).await?
+            } else if reward_epoch_duration + sleep_duration >= reward_period {
+                // If the next epoch is a reward period, cut off sleep duration.
+                // This ensures that verifying will always end up being aligned with
+                // the desired reward period.
+                sleep_duration = reward_period - reward_epoch_duration;
+            }
+
+            tracing::info!("Sleeping...");
+            let shutdown = shutdown.clone();
+            tokio::select! {
+                _ = shutdown => return Ok(()),
+                _ = sleep(
+                    sleep_duration
+                        .to_std()
+                        .map_err(|_| Error::OutOfRangeError)?,
+                ) => (),
+            }
+        }
+    }
+
+    pub async fn verify_epoch(&mut self, epoch: Range<DateTime<Utc>>) -> Result {
+        let transaction = self.pool.begin().await?;
+
+        let shares: Shares = self.verifier.verify_epoch(&epoch).await?;
+
+        // Should we remove the heartbeats that were not new
+        // from valid shares
+        for share in shares.valid_shares.clone() {
+            let heartbeat = Heartbeat::from(share);
+            heartbeat.save(&self.pool).await?;
+        }
```

Comment:
if you're doing these all at once why not do a bulk insert or insert in batches?

Reply:
I don't think that's actually possible with sqlx: 
https://github.com/launchbadge/sqlx/issues/294

Reply:
I think this is different... update_all is updating all entries with a given set of ids to the same timestamp. We don't necessarily want to update all of these entries to the same values, so in order to use this method, we'd have to group up our entries that are being set to the same value, which kinda just becomes the same thing we're doing now but with more pre-processing

Reply:
granted, that's a simplistic case b/c you're bulk updating the records that match the `any` clause with the same values, but skimming that issue it looks like there are a couple of ways sqlx will let you do it that aren't even very hacky anymore (UNNEST, ToRow, QueryBuilder); seems like a good opportunity to throw a `TODO` in here and revisit it when the thing is up and running?

Reply:
Yeah, a todo would make sense here